### PR TITLE
feat: excel import

### LINKED ISSUE #502: Improve excel import
## The issue
Currently, when the user imports an excel file, we go through a custom parsing step:
- Only worksheets with certain names are considered by ontime
- We look for columns with certain names to match to ontime values

The function is ugly but works well. It is easy to extend and safely ignores things it doesn't understand
https://github.com/cpvalente/ontime/blob/master/apps/server/src/utils/parser.ts#L39

However, this makes it inflexible for user who receive rundown or event spreadsheets from clients at the last minute and have to cleanup for Ontime import

## The task
The overall task is to create an intermediate step between the user uploading an excel file and it being parsed into the rundown.
This will likely entail:

### Create preParsing() function

It should receive as arguments: 
- a worksheet name
- an object where the key-value corresponds to `event object fields` - `user given column name`, eg: `"timerStart": "Event Start"`

and return a parsed rundown array for the user to confirm

### Create the necessary UI
We would need to add a few steps to post-upload. I imagine this would be part of the existing Upload component. Although it admittedly means some large changes (see ugly screenshot)
- Form where user inputs the worksheet and column names
- Table that shows the parsed rundown and allows user to confirm import

import

### Tie it all together
To get the feature complete, we would need to
- create an endpoint `resolveExcel` that receives the values of `preParsing` function and returns its result
- create an endpoint `postRundown` that receives a rundown and updates the app data

The overall steps would be something like
- user opens upload component
- user uploads file
- if file is an excel file we show the form for field input 
- on submit to `resolveExcel`, we run the data through `preParsing` and return the result
- display the result in a table
- A new "Confirm Import" button pushes the rundown object to the new endpoint `postRundown`

### LINKED ISSUE #508: Resolve times on excel import
## The issue
Currently, when importing from excel, we take an event `start time` and `end time` as given by the user and calculate the duration as derived.

We would like to a) give the user the possibility to add partial data b) ensure the data provided is correct before entering the app

## The task

The issue task is two-part:
1. On parsing an excel datasheet we need to read the extra field duration
2. On the parseRundown step, we need to validate the time data so that:
- If **all given**: `duration = end - start`
- if **only duration**: `start = 0; end = duration`
- if **only start** is given: `start = duration = 0`
- if **start + duration** is given: `end = start + duration`
- if **only end** is given: `start = 0; duration = end` 
- if **end + duration** is given: `start = end - duration`
- if **start + end** is given: `duration = end - start`

This task should either follow or be implemented as part of #502

### PR DESCRIPTION
**Pull request by cpvalente**
adds review step to excel import with user-given import map for the column titles, closes #502 

Also closes #508

### CODE REVIEW

File: apps/server/src/utils/parser.ts [hunk 1]
```diff
@@ -119,184 +156,158 @@ export const parseExcel = async (excelData) => {
...
+                user8Index = j;
+                break;
+              case importMap.user9:
+                user9Index = j;
+                break;
+
+              default:
+              // we don't know how to handle this column
+              // just ignore it
             }
```
[the remark this answers is not in the document]
Comment (cpvalente) [reply apps/server/src/utils/parser.ts:283]: can you provide a code example?

**Comment by alex-Arc** [comment]
If a file without project data fields is uploaded the project data already in the show gets overridden to blank
Is that an intentional thing?

File: apps/client/src/features/modals/upload-modal/preview/PreviewTable.module.scss [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  tr {
+    word-wrap: unset;
+  }
+}
+
+.body {
+  td {
+    text-align: left;
```
Comment (alex-Arc) [inline apps/client/src/features/modals/upload-modal/preview/PreviewTable.module.scss:45]: the preview could use some line sepperation
image
|>
image

**Comment by cpvalente** [comment]
> If a file without project data fields is uploaded the project data already in the show gets overridden to blank
Is that an intentional thing?



I think so, when the user imports a project file we should override everything. From the users perspective they started a new project

**Pull request merged**